ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerChannelImpl.java

```diff
@@ -213,39 +224,61 @@ public SafeFuture<TransitionConfiguration> engineExchangeTransitionConfiguration
   }
 
   @Override
-  public SafeFuture<ExecutionPayloadHeader> getPayloadHeader(
+  public SafeFuture<ExecutionPayloadHeader> builderGetHeader(
       final Bytes8 payloadId, final UInt64 slot) {
-    LOG.trace("calling getPayloadHeader(payloadId={}, slot={})", payloadId, slot);
+    checkState(
+        lastSentForkChoiceState.isPresent()
+            && lastSentForkChoiceState.get().getHeadBlockSlot().equals(slot));
+    checkState(executionBuilderClient.isPresent() && lastSentPayloadAttributes.isPresent());
 
-    return executionEngineClient
-        .getPayloadHeader(payloadId)
+    LOG.trace("calling builder builderGetHeader(payloadId={}, slot={})", payloadId, slot);
+
+    return executionBuilderClient
+        .get()
+        .getHeader(slot, Bytes48.ZERO, lastSentForkChoiceState.get().getHeadExecutionBlockHash())
         .thenApply(ExecutionLayerChannelImpl::unwrapResponseOrThrow)
-        .thenCombine(
-            SafeFuture.of(
-                () ->
-                    SchemaDefinitionsBellatrix.required(spec.atSlot(slot).getSchemaDefinitions())
-                        .getExecutionPayloadHeaderSchema()),
-            ExecutionPayloadHeaderV1::asInternalExecutionPayloadHeader)
+        .thenApply(
+            builderBidV1SignedMessage ->
+                getExecutionHeaderFromBuilderBid(builderBidV1SignedMessage, slot))
         .thenPeek(
             executionPayloadHeader ->
                 LOG.trace(
-                    "getPayloadHeader(payloadId={}, slot={}) -> {}",
+                    "builderGetHeader(payloadId={}, slot={}) -> {}",
                     payloadId,
                     slot,
                     executionPayloadHeader));
   }
 
+  private ExecutionPayloadHeader getExecutionHeaderFromBuilderBid(
+      SignedMessage<BuilderBidV1> signedBuilderBid, UInt64 slot) {
+    ExecutionPayloadHeaderSchema executionPayloadHeaderSchema =
+        SchemaDefinitionsBellatrix.required(spec.atSlot(slot).getSchemaDefinitions())
+            .getExecutionPayloadHeaderSchema();
+    // validate signature
```

Comment:
Where does the public key to validate the signature come from? If it's a validator key then it would have to come from state and we won't be able to verify the signature over here - we'd have to pass it all back to the beacon node unchecked which is messy...

Reply:
Seems like we should make sure we have a ticket to come back and actually validate the signature here.